app/Libraries/Payments/PaypalCreatePayment.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+/**
+ *    Copyright 2015-2017 ppy Pty. Ltd.
+ *
+ *    This file is part of osu!web. osu!web is distributed with the hope of
+ *    attracting more community contributions to the core ecosystem of osu!.
+ *
+ *    osu!web is free software: you can redistribute it and/or modify
+ *    it under the terms of the Affero GNU General Public License version 3
+ *    as published by the Free Software Foundation.
+ *
+ *    osu!web is distributed WITHOUT ANY WARRANTY; without even the implied
+ *    warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ *    See the GNU Affero General Public License for more details.
+ *
+ *    You should have received a copy of the GNU Affero General Public License
+ *    along with osu!web.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+namespace App\Libraries\Payments;
+
+use App\Models\Store\Order;
+use App\Traits\StoreNotifiable;
+use PayPal\Api\Amount;
+use PayPal\Api\Details;
+use PayPal\Api\Item;
+use PayPal\Api\ItemList;
+use PayPal\Api\Payer;
+use PayPal\Api\Payment;
+use PayPal\Api\RedirectUrls;
+use PayPal\Api\Transaction;
+use PayPal\Exception\PayPalConnectionException;
+
+/**
+ * Creates a Paypal Payment for a store Order.
+ *
+ * Using the Paypal REST API for payments is a 2-phase process;
+ * This class handles the creation of a payment for the user to approve.
+ */
+class PaypalCreatePayment
+{
+    use StoreNotifiable;
+
+    private $order;
+    private $payment;
+
+    public function __construct(Order $order)
+    {
+        $this->order = $order;
+
+        $payer = new Payer();
+        $payer->setPaymentMethod('paypal');
+
+        $payment = new Payment();
+        $payment->setIntent('sale')
+            ->setPayer($payer)
+            ->setRedirectUrls($this->getRedirectUrls())
+            ->setTransactions([$this->getTransaction()]);
+
+        $this->payment = $payment;
+    }
+
+    public function getApprovalLink()
+    {
+        $context = PaypalApiContext::get();
+
+        try {
+            $this->payment->create($context);
+
+            return $this->payment->getApprovalLink();
+        } catch (PayPalConnectionException $e) {
+            \Log::error($e->getData());
+            // TODO: get more context data
+            $this->notifyError($e, $this->order);
+            throw $e;
+        }
+    }
+
+    public function getPayment()
+    {
+        return $this->payment;
+    }
+
+    private function getAmount()
+    {
+        $details = new Details();
```

Comment:
looks like a bunch of this assignment/new + set/add (+ return) can be done in single step because set/add returns `$this`.